### DISCUSSION TITLE
fp8 GGUF and GPU accelerated inference support

### DISCUSSION THREAD

**Discussion by jim-plus** [ggml-org/llama.cpp#8780] [Ideas]
How difficult would it be to support conversion to fp8 for GGUF, and to add accelerated GPU support? I have a 4060ti 16gb Lovelace GPU and am interested in leveraging its fp8 support.

**Comment by charyang-ai** [comment]
@jim-plus  is there any progress on the FP8 support? I am working on AMD GPU and also need this FP8 feature.

**Comment by teleprint-me** [comment]
It's straightforward enough, but the accuracy is worse than int8.

```c
typedef union {
    float value; /**< Floating-point value */
    uint32_t bits; /**< Raw bit representation */
} Float32;

// Encode a float into an 8-bit floating-point representation
uint8_t encode_float8(float value) {
    if (value == 0.0f) {
        return 0; // Encoded as all zeros
    }

    Float32 encoder = {.value = value};

    // Extract IEEE-754 components
    uint32_t sign = (encoder.bits >> 31) & 0x1;
    uint32_t exponent = (encoder.bits >> 23) & 0xff;
    uint32_t mantissa = encoder.bits & 0x7fffff;

    // Define bias parameters
    uint32_t e_bias_32 = 127;
    uint32_t e_bias_8 = 3;

    // Define exponent limits
    uint32_t e_max = 7;
    uint32_t e_min = 0;

    // Calculate compressed exponent
    int8_t e_compressed = fmaxf(fminf(exponent - e_bias_32 + e_bias_8, e_max), e_min);

    // Calculate compressed mantissa (top 4 bits of the 23-bit mantissa)
    uint8_t m_compressed = (mantissa >> 19) & 0xf;

    // Pack into an 8-bit integer
    return (uint8_t) ((sign << 7) | (e_compressed << 4) | m_compressed);
}

// Decode an 8-bit floating-point representation back to a float
float decode_float8(uint8_t bits) {
    // Extract fields
    uint8_t sign = (bits >> 7) & 0x01;
    uint8_t exponent = (bits >> 4) & 0x07;
    uint8_t mantissa = bits & 0x0F;

    // Define parameters
    uint32_t e_bias_32 = 127;
    uint32_t e_bias_8 = 3;

    // Expand exponent
    int32_t e_expanded = exponent - e_bias_8 + e_bias_32;

    // Expand mantissa with implicit leading 1
    float m_expanded = 1.0f + (mantissa / 16.0f);

    // Reconstruct float
    float result = ldexpf(m_expanded, e_expanded - e_bias_32);
    return sign ? -result : result;
}
```

I've run multiple experiments between int8 and fp8 and int8 always reduces the margin of error. int8 is also much faster.

**Reply by teleprint-me** [reply]
As an added note, I would avoid ROCm. There's too many issues with it and the developers always deprecate cards as fast as they update the API. Vulkan would give you the desired 8-bit support your looking for as an integer format. It's what I usually use since I have a similar model GPU. e.g. It's already implemented.

**Reply by netrunnereve** [reply]
> It's straightforward enough, but the accuracy is worse than int8.

Do you mean that it's worse than Q8_0 or that it's worse than unscaled int8? FP8 is intended to be used with a floating point block scale like Q8_0 and Nvidia uses that approach in their own libraries.

I think the way to think of FP8 vs int8 in llama.cpp terms is to basically compare iq4_nl against Q4_0. Both have a scaling factor but the dynamic range of the quantized weights is increased, creating a perplexity improvement over Q4_0. While iq4_nl requires a LUT newer hardware supports FP8 natively and performance there should be pretty similar to int8.

> Vulkan would give you the desired 8-bit support your looking for as an integer format. It's what I usually use since I have a similar model GPU. e.g. It's already implemented.

Our Vulkan implementation currently does all calculations in FP16 or FP32 and doesn't quantize the activations to Q8_0 like CPU or CUDA. We don't use this but Vulkan has the VK_KHR_shader_integer_dot_product extension if someone wants to implement an int8 based path in the future.

BTW someone is working on adding FP8 support for CPU [here](https://github.com/ggerganov/llama.cpp/pull/10055).

**Reply by Djip007** [reply]
Yes there is a PR... I need to rebase on last ggml change...

Note: your compute float->FP8 is wrong, you trunc the mantice not rounding ... and it make real difference to .
What I find unfortunate is that out of the many possible formats to implement FP8 (E5M2, E4M3, E3M4 ...) it is not the best according to my tests that is implemented in the hardware. In my tests the most efficient are the E3M4. this is what happens when we do not check the results of research papers before applying them.

Next I have really close result with E4M3_Q / E3M4_Q and Q8_0, but Q8_0 need 1 scale for 32 value, and we can use an scale for 256/512 FP8. 

with mistral_nemo: (ref: BF16) (1 scale for 256 fp8)
|   BF16   |    PPL   |    KLD   | top P   |
| -------: | -------: | -------: | ------: |
|   FP16   | 6.339475 | 0.000002 |  99.941 |
|   Q8_0   | 6.345049 | 0.000854 |  98.503 |
|  E3M4_Q  | 6.345332 | 0.001324 |  98.045 |
|  E4M3_Q  | 6.361256 | 0.004980 |  96.182 |
|   Q6_K   | 6.373544 | 0.006197 |  95.778 |
|  Q5_K_M  | 6.397938 | 0.010158 |  94.834 |      
|  Q5_K_S  | 6.437732 | 0.013509 |  94.231 |      
|  Q4_K_M  | 6.489127 | 0.026584 |  92.176 |

**Reply by teleprint-me** [reply]
@netrunnereve 

I meant scaled int8, unscaled int8 is awful. You lose too much precision. Q8_0 uses unscaled, but it is mapped to the integer domain based on intervals of the block size.

The samples are based on calculated strides. The delta is calculated based on every kth sample based off of the stride, then the inverse is multiplied by the element of the set. The scalar is stored as half.

According @0cc4m, int8 (Q8_x) is already implemented. 

I haven't looked at the code in the past 2 weeks, so it's possible I missed something. The codebase massive. I can't recall every detail on demand.

@Djip007 

You're right, I did not round which does affect the precision.

My goal was to keep the initial implementation simple. I got lazy, but I shared what I have at the moment.

I have looked at the math, research, and application behind 8-bit applications (e.g. dsp, dettmers, etc.), but have have not had the opportunity to look into fp8 in depth. It was an initial prototype I sketched out, but never fleshed out.

There are other methods with varying degrees of success as well. I haven't finetuned my approach with fp8 as much as I have with int8.

The issue with fp8 is that it assigns bits to each region of the total bit width and you always lose too much information as a result due to a much smaller domain. Might as well use a packed int8 (Q4) at that point.

With int8 (Q8), the issue is just scaling the real to the integer domain.

**Reply by Djip007** [reply]
some more result here: https://github.com/Mozilla-Ocho/llamafile/discussions/549#discussioncomment-10988233

**Reply by netrunnereve** [reply]
> Q8_0 uses unscaled, but it is mapped to the integer domain based on intervals of the block size.

I'm a bit confused here, as Q8_0 is a scaled format that multiplies a int8 value with a float scale (or delta depending on how you call it) to get the final floating point weight. Nvidia does the same thing but they use a fp8 instead of an int8 value.

> The samples are based on calculated strides. The delta is calculated based on every kth sample based off of the stride, then the inverse is multiplied by the element of the set. The scalar is stored as half.

This is how you actually quantize the original floating point numbers into Q8_0, right? I'm not really into the mathematical side of this and I remember almost flunking Linear Algebra back in my undergrad days :cry:.

> According @0cc4m, int8 (Q8_x) is already implemented.

Oh you meant Q8_0 support, yeah that has been there for a long time. However even if the GPU supports native int8 math our implementation doesn't use it. In the future I'm hoping that someone will go write those quantized mat mul routines for Vulkan just like what we have for CUDA.
```
vec4 dequantize4(uint ib, uint iqs, uint a_offset) {
    const float d = float(data_a_packed16[a_offset + ib].d);
    uint32_t v0 = data_a_packed16[a_offset + ib].qs[iqs/2];
    uint32_t v1 = data_a_packed16[a_offset + ib].qs[iqs/2 + 1];
    return vec4(int8_t(v0 & 0xFF), int8_t((v0 >> 8) & 0xFF), int8_t(v1 & 0xFF), int8_t((v1 >> 8) & 0xFF)) * d;
}
```

**Reply by teleprint-me** [reply]
@Djip007 

I'm Looking into it. [NVidia](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/examples/fp8_primer.html) has a good break down here. Helped me parse your initial opening comment.

@netrunnereve

It's cool, I can break it down in simple terms. I failed out of high school and have been teaching my self the math over the past decade since I can't afford to go back to school. It's actually really simple once you get the hang of it. Takes a bit to understand, though.

We have natural, integer, real, etc. numbers. So the $Z$ domain is a **set** that represents the **integer space**.

```python
Z = {0, 1, 2, 4, 5} # integers going on infinitely, but this keeps it simple
```

$Z$ becomes our set with elements 1 through 5. A set contains no repeating elements. We have 8-bits, so that's a base of 2 with a power of 8, `pow(2, 8)`. This just means we can "fit" any number in the range between 0 and 255 inclusive. This is the **integer domain**.

```python
>>> (2**8) - 1
255
```

$R$ becomes our set with random elements. We can call this our **sampled set**. Samples are just probabilities within a given space. For example, "What are the odds of a value occurring in the real space?" (this is a trick question). A real number is just a fractional representation of a number. For example, `1.23456` is a real number. Any number that fits this definition on a number line is in the **real domain**.

An element in a domain is simple enough. We say that $z \in Z$ is read as "z is in the set of integers". So, if 0 is in Z, then the statement is true.

```python
>>> 0 in Z
True
```

Let say we have a sampler that generates real numbers and they fall within a 16-bit range. This gives us a range of about 65535 possible states.

```python
>>> (2**16) - 1
65535
```

This is obviously never going to fit inside an 8-bit space. It's not physically possible. We have to apply a few tricks to make this work. The easiest trick is to just get the **step size**. In a 8-bit space, we can say the smallest step is 1 which is just $\frac{1}{255}$ - this just means we took $1$ step out of $255$ possible steps.

Making a basic sampler is really simple.

```python
def sampler(max_elements: int, z: int) -> list[float]:
    assert max_elements > 0
    assert z >= 2
    values = []
    for i in range(max_elements):
        # Generate a random value in the range [0, 1]
        normalized: float = random()
        # Map the normalized value to the range [-n, n-1]
        values.append(-z + (normalized * (2 * z - 1)))
    return values
```

This just takes in the number of elements you expect the list to have and the defined range of that list. This means any generated number will always be within whatever range $z$ is.

```python
seed(1337) # make the random number generator predictable
values = sampler(10, 2**8 - 1) # create 10 elements in the range [-255, 255]
```

If we want some samples for the real space, we can just do the following.

```python
>>> samples = sampler(5, 10)
>>> samples
[-2.2997372367656492, -5.7918604002338165, 1.280326098515765, 0.5747578675907388, -2.9595928979118638]
>>> 
```

So, we just sampled $5$ **real values** in the range $[-10, 9]$. Each value generated is a **sample**. We can change our step size from $1$ to anything.

If our step size is $1$, we go through all of the elements we have.

If we increase the step size to $2$, we do every other element in the set.

The **step size** can also be called the **frequency** or **sample rate**.

The quant functions are just linear mappings. This just means for every input, we expect a predictable output in a given dimension. We can say the dimension is $1$ which is just the number line. We usually call the horizontal line in a graph $x$.

```python
def f(x):
    return 2 ** x
```

$f(x)$ just maps $x$ to the exponent with a base of $2$ as $y$ where $y$ is the output of the function $f$.

Now that we have the boring stuff out of the way, we can get to fun part.

LLM's have really large sets with varying shapes. Usually $1$ dimension (a vector) or $2$ dimensions (a matrix). These are the **shapes**. There's a bunch of real numbers (aka floating-point) in there that the function "learns".

We have no idea how or why the mappings occur the way they do (the area I'm attempting to specialize in), but we do know that it's computationally expensive with the amount of numbers in these sets as they grow - exponentially or even quadratically.

We'll use a small sample set of possible values that occur in any model, since they're usually small.

```python
>>> samples[0:len(samples)-1:2] # 0 is start, len(samples-1) is stop, 2 is step
[-2.2997372367656492, 1.280326098515765] # positions 0 and 2
```

Here, we just grab the elements at 0 and 2. 1, 3, and 4 are ignored.

We want to take these numbers and fit them into a smaller space. Right now, they're 32-bits. This gives us $4294967295$ possible values.

```python
>>> 2**32-1
4294967295
```

We need to fit this in a smaller space. 8-bits. This is not easy to do, but it's usually just because it breaks the way we usually think about it.

We need to stuff the sampled values into a smaller space. 32-bit to 8-bit. We also need to somehow remember how to get the original value back. So we create a linear function to do this for us.

$g(x) : x \to y$ can be read as $x$ maps to $y$. $x$ is our floating point and $y$ is our quantized value representing the original value. Since $y$ maps to $x$, we should be able to get back to $x$ from $y$.

So, we put our value in and get some integer out.

```sh
00:22:00 | /mnt/valerie/private/alt
 git:(main | Δ) λ python examples/quants/q8.py 
Testing Quantization and Dequantization:
Input: 59.436204, Quantized: 128, Scalar: 0.466166, Alpha: 1.000000, Residual: -0.233083, Reconstructed: 59.436204
```

Here, $x = 59.436204$ and $y = 128$. We scaled $x$ down to fit inside of the range we defined for 8-bits. But we need to get it back somehow. We can scale back up. We say this is the approximate (prime, image, etc) of the number. The reconstructed value is $\hat{x} = 59.436204$.

The math to do this is super simple.

```python
# Calculate the base step size
scalar = r_domain / z_domain
```

We just divide the domains by one another. We want the fraction to go into the integer. To get the original number back, we just do the opposite.

```python
q8.bits * q8.scalar
```

What GGML does is it takes a block of 32 elements while skipping steps (just like hopscotch) and shrinks every value it finds to fit it into a smaller space.

The reason for doing this is obvious to most of us as users. It's faster and uses less space.

Sorry this was so long! It got out of hand, lol.

**Reply by Djip007** [reply]
> Nvidia does the same thing but they use a fp8 instead of an int8 value.

@netrunnereve 
Did you find what Nvidia really did with FP8? (can't find clear "paper"):
Did they use per bloc scale, per row or global tensor? [Llama-3.1-405B-Instruct-FP8](https://huggingface.co/meta-llama/Llama-3.1-405B-Instruct-FP8/tree/main?show_file_info=model-00004-of-00109.safetensors) in that case it is per row scale but it is not use on all matmul weight.

**Reply by netrunnereve** [reply]
This is a very clear explanation, thank you!

What I was thinking about fp8 comes from the fact that that most LLM weights are small numbers - I think the majority are in the -1 to 1 range and are distributed around 0. If within the block of 32 numbers we have outliers then the range may end up being [-10, 10] with 90% of the numbers between [-1, 1]. Due to the large range we will have to use a large step size of `20/256 = 0.08` when quantizing to int8. With this high step size small numbers won't be represented as accurately, and we would poorly utilize our 8 bits as only 25 steps out of 256 represent our [-1, 1] region.

To help with this fp8 provides a dynamic step size. The outliers are handled by the edges of the fp8 domain while the [-1, 1] range is handled with a lot more precision.

[FP8 tables for reference](https://asawicki.info/articles/fp8_tables.php)

Well I wrote all that up and I realized that this wouldn't work if the block real domain was something like [-1, 10] as all the fp8 accuracy would be concentrated around the 4.5 mark. Maybe fp8 would make more sense with something like Q4_1 as the additional offset could help push the smallest step size region towards where most of the weights are clustered.

**Reply by Djip007** [reply]
Hop they add E3M4 FP8 format (https://github.com/IntelLabs/FP8-Emulation-Toolkit).

We can have different scale compute, for my case I simply use scale so for a bloc:

`A[i] = A.fp8[i]*A.scale[i/K]`
 the EXMY_Q use K=256 for now (I test many more from 1024 <> 128) 
to compute the scale I have apply a simple patern, juste get the `max(A[i:i+K])` and compute the scale so `max(A.fp8[i,i+K]) = fp8_Max`.

Use QN_1 like quant have some drawback: it look like: `A[i] = A.fp8[i]*A.scale[i/K]+A.bias[i/K]`
with QN_0 on dot_product you can apply the scale after K madd. (simple factorization)
But with QN_1 bias have to be apply before madd. (Yes some CPU can do 1 fma + 1 add per cycle... but...)

With the height range we have with fp8 (compare to int8) and the fact we do not need Q8_1 for good result, I assume it may be overkill to use it with fp8...

I wait for my last perplexity compute, but look my last change made some case even better... more in 2h ;) 

Note: I assume it is the same for int8 all bloc of 32 int8 have 1 value a 255 ou -256...

**Reply by netrunnereve** [reply]
@Djip007 

> Did you find what Nvidia really did with FP8? (can't find clear "paper"):
Did they use per bloc scale, per row or global tensor? [Llama-3.1-405B-Instruct-FP8](https://huggingface.co/meta-llama/Llama-3.1-405B-Instruct-FP8/tree/main?show_file_info=model-00004-of-00109.safetensors) in that case it is per row scale but it is not use on all matmul weight.

They have some docs [here](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/examples/fp8_primer.html) and [here](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/api/common.html#transformer_engine.common.recipe.DelayedScaling) for their transformer library which indicates that by default they compute one scale every 1024 values for *training*. I would imagine that this value is intended to be tweakable in order to balance perplexity and model size.

Honestly I just know that they require a scale with fp8, and I'm not sure about their exact implementation.

@teleprint-me

And I just found [this paper](https://arxiv.org/pdf/2303.17951) which compares int8 and fp8 in detail.

*"This leads us to a very simple conclusion: It’s all about the outliers. If a distribution has very signifi-
cant outliers, the FP8-E4/FP8-E5 format is more accurate, but if the distributions are well-behaved
and more Gaussian-shaped, then INT8 or FP8-E2/FP8-E3 is expected to perform better. This is the
very simple idea that explains much of the behavior of these formats in practice, as we will see in the
following sections.
It is important to note that adding exponent bits in favor of mantissa bits can hurt accuracy significantly.
For well-behaved Gaussian layers, the FP8-E5 format only gives around 4 or 5 bits of information.
This means that if the scale factor is set correctly, you could have even used a INT4/INT5 format
instead to achieve the same accuracy.
Lastly, even in the ideal scenario for distributions with outliers, the bit-accuracy is lower than for
distributions without outliers. You can see this in Figure 5 at the bottom. The highest score for the
uniform distribution is 8 bits, whereas for the student-t-distribution FP8-E4 has only 5 bits of accuracy.
Thus, to optimize your network for quantization, it is best to have a network without any outliers at
all. For such a network, the INT8 and FP8-E2 formats would perform best. This will comes back in
Section 4.5 on QAT. In training networks for quantization, it is easy to train outliers away, resulting in
better results overall, and in many cases, the INT8 format then outperforms the floating-point formats."*

**Reply by Djip007** [reply]
Some last number (my laptop needs a rest), with rounding to nearest even / mistral nemo.

|   BF16   |    PPL    |    KLD   | top P   |
| -------: | --------: | -------: | ------: |
|   FP16   |  6.339475 | 0.000002 |  99.941 |
|   Q8_0   |  6.345049 | 0.000854 |  98.503 |
|  E3M4_Q  |  6.345332 | 0.001324 |  98.045 |
| **E3M4_C**  | **6.346416** | **0.001562** | **97.923** |
|  E4M3_Q  |  6.361256 | 0.004980 |  96.182 |
|   Q6_K   |  6.373544 | 0.006197 |  95.778 |
|  E4M3_C  |  6.374483 | 0.006227 |  95.598 |
|  Q5_K_M  |  6.397938 | 0.010158 |  94.834 |
|  Q5_K_S  |  6.437732 | 0.013509 |  94.231 |
|  E5M2_Q  |  6.452403 | 0.018773 |  92.824 |
|  E5M2_C  |  6.493300 | 0.022102 |  92.156 |
|  Q4_K_M  |  6.489127 | 0.026584 |  92.176 |
|  E5M2_G  |  6.511753 | 0.027467 |  91.472 |
|  E4M3_G  | 34.541798 | 1.738122 |  49.099 |

Q: 1 scale for 256 fp8
C:  1 scale par row
G: 1 scale per tensor

As you can see (it was not as good with other rounding) even per line scale is good ... and with it no need to apply scale in the matmul compute. it is impossible with int8...
per tensor scale only work for E5M2 but it is expected as it have the same Exponent as FP16 (event if nemo is distributed with BF16...)

**Reply by Djip007** [reply]
> And I just found [this paper](https://arxiv.org/pdf/2303.17951) which compares int8 and fp8 in detail.

https://github.com/quic/aimet/issues/2678

Nice paper.

But, most study in this paper is with small model and mostly CNN ... they have not test large LLM, and we know that there per-tensor/per-channel won't work with INT8... the BERT that is the most close to ours LLM (bet really smal) is the one who is the exception in their paper :sunglasses: 
And as you see FP8 can (in some model) have per-row quantisation with good result... 

LLM have large matmul OP, that "love" Floating point computation...

**Reply by teleprint-me** [reply]
Yes, it's because GGML does not account for the residual (bias) or alpha (squeezing factor). So, I'm not surprised that FP8 is close to INT8. In fact, it's aligned with the math. A few tweaks to the Q8 algorithm will change that dramatically.

I'm not saying it's not worth investigating and experimenting with FP8, but I would not personally advocate for FP8. If you understand how the math works, how the hardware works, and why integers are more efficient and precise, it becomes obvious.

I can totally see this becoming like a Vim vs Emacs thing, lol. But unlike Vim and Emacs, the math isn't as subjective.

Here's something you can do without testing, just with pen and paper. Compare the range of each domain for FP8 and INT8. Which uses space more efficiently?

**Reply by teleprint-me** [reply]
@netrunnereve

You might find this interesting and helpful.

[Applied Digital Signal Processing](https://www.youtube.com/watch?v=yGeXEwdNd_s&list=PL_QS1A2ZqaG7p50cd0AgLeG9Q3TN64vZJ)

It helped clarify the concepts for me. I've always been enamored with music, so the audio concepts really helped visualize the math and gave it a more concrete and grounded form of intuition.

Episodes 3, 4, 5 really drive home the concepts we've been discussing here.

**Reply by Djip007** [reply]
> Yes, it's because GGML does not account for the residual (bias) or alpha (squeezing factor).

I don't know what you think bias/squeezing factor are, but it is widely used in ggml.
```cpp
#define QK8_0 32
typedef struct {
    ggml_half d;       // delta
    int8_t  qs[QK8_0]; // quants
} block_q8_0;

#define QK8_1 32
typedef struct {
    union {
        struct {
            ggml_half d; // delta
            ggml_half s; // d * sum(qs[i])
        };
        ggml_half2 ds;
    };
    int8_t qs[QK8_1]; // quants
} block_q8_1;
```
Q8_0 quant use scale,
Q8_1 quant use bias+scale. **Correction: not for this** 
it is the case for other QN_1:
```c++
#define QK4_1 32
typedef struct {
    union {
        struct {
            ggml_half d; // delta
            ggml_half m; // min
        };
        ggml_half2 dm;
    };
    uint8_t qs[QK4_1 / 2]; // nibbles / quants
} block_q4_1;
```

Now keep in mind that most quantize tensor is use in large matmul that result on for exemple on Mistral_Nemo with dot product on vector of size 14336 for the longer.

**Reply by teleprint-me** [reply]
It's not complicated.

```c
// reference implementation for deterministic creation of model files
void quantize_row_q8_0_ref(const float * restrict x, block_q8_0 * restrict y, int64_t k) {
    assert(k % QK8_0 == 0); // assert even divisible by the block size
    const int nb = k / QK8_0; // calculate the stride

    for (int i = 0; i < nb; i++) {
        float amax = 0.0f; // absolute max

        for (int j = 0; j < QK8_0; j++) {
            const float v = x[i*QK8_0 + j];
            amax = MAX(amax, fabsf(v)); // get the absolute max from the block
        }

        const float d = amax / ((1 << 7) - 1); // calculate the integer domain
        const float id = d ? 1.0f/d : 0.0f; // creates an inverse to fit the domain

        y[i].d = GGML_FP32_TO_FP16(d); // packs the scalar into 16-bits

        for (int j = 0; j < QK8_0; ++j) {
            const float x0 = x[i*QK8_0 + j]*id; // scales the float down to integer

            y[i].qs[j] = roundf(x0); // sets the quantized value
        }
    }
}
```

Not once does it account for the bias or the squeezing factor.

FP8 is not capable of accounting for the squeezing factor. You can only do this with an INT8.

If you want more details on what I mean by this, you can look [here](https://github.com/teleprint-me/alt.c/blob/main/docs/quantization/error-free-modeling.md).

The math proves that the scaling and bias are insufficient for properly approximating the original value. That's why another variable is required.

FP8 is hobbled by allocating bits to the sign, exponent, and mantissa. But you can "bake" that information into an INT8. This increases the range of the integer domain as a result. This in turn outperforms FP8.

If you look at DSP methods, it becomes clear why they use 16-bit and can reduce so much noise from the output signal. That's why the input values are normalized to $[-1, 1]$. It normalizes the range proportionally from 32-bits to 16-bits.

The same concept can be applied to any other range, but introduces a higher standard deviation as a result. That's why you can hear the noise in audio output signals as the bit-width decreases.

**Reply by Djip007** [reply]
```c++
void quantize_row_q4_1_ref(const float * restrict x, block_q4_1 * restrict y, int64_t k) {
    const int qk = QK4_1;

    assert(k % qk == 0);

    const int nb = k / qk;

    for (int i = 0; i < nb; i++) {
        float min = FLT_MAX;
        float max = -FLT_MAX;

        for (int j = 0; j < qk; j++) {
            const float v = x[i*qk + j];

            if (v < min) min = v;
            if (v > max) max = v;
        }

        const float d  = (max - min) / ((1 << 4) - 1);
        // - max-min is what you call real domain
        // - (1 << 4) - 1=15  is what you call integer domaine
        // => d is what Squeezing Factor
        const float id = d ? 1.0f/d : 0.0f;

        y[i].d = GGML_FP32_TO_FP16(d);
        y[i].m = GGML_FP32_TO_FP16(min);  // this is the bias! (that is not define in you link)

        for (int j = 0; j < qk/2; ++j) {
            const float x0 = (x[i*qk + 0    + j] - min)*id;
            const float x1 = (x[i*qk + qk/2 + j] - min)*id;

            const uint8_t xi0 = MIN(15, (int8_t)(x0 + 0.5f));
            const uint8_t xi1 = MIN(15, (int8_t)(x1 + 0.5f));

            y[i].qs[j]  = xi0;
            y[i].qs[j] |= xi1 << 4;
        }
    }
}
```
(and yes look like Q8_1 do not compute the bias.)

> FP8 is not capable of accounting for the squeezing factor. You can only do this with an INT8.

I don't know why you think it is not possible.
```c++
template <int E, int QK>
static inline void conv(const float* x, bloc_fp8<E, QK>* y, int64_t size) {
    const auto qk_size = size / QK;
    for (int64_t q=0; q<qk_size; ++q) {
        float m = 0;
        for (int64_t i=0; i<QK; i++) {
            m = std::max(std::abs(x[q*QK+i]),m); 
        }
        const float D = FP8<E>::MAX()/m;
        y[q].d = m/FP8<E>::MAX();  // what you call squeezing factor
        for (int64_t i=0; i<QK; i++) {
            y[q].qs[i] = x[q*QK+i]*D;
        }
    }
}
```

**Reply by teleprint-me** [reply]
Here are the sources I used to derive the math. I also worked through the applications with GPT while providing GPT the necessary information and context.

- https://www.math.net/domain-and-range
- https://discrete.openmathbooks.org/dmoi3/sec_intro-sets.html
- https://discrete.openmathbooks.org/dmoi3/sec_intro-functions.html
- https://en.wikipedia.org/wiki/Domain_of_a_function

If I use 3 bits for the exponent and 4 bits for the mantissa, I've immediately compromised half the bit width and reserved a bit for a single sign.

This creates a domain of $2^4$ bits. That's why FP8 is capped at around ~ $\pm 450$. It's a balancing act between range and precision. The upside is you use a single byte at the cost of increased complexity (both in hardware and application). This last part is the efficiency loss. The gain is the reconstruction process is fairly straightforward and requires no extra metadata. This cost puts in parity with a Q4 while operating at the near precision of Q8 depending on the context.

With INT8, you get the full bit width, but at the cost of requiring scalar and sign. This leaves you with about $2^7$ bits. The scalar adds an extra 2-bytes, but allows you to store away metadata required for reconstruction. Reconstruction is fairly simple and arithmetic is fairly cheap in comparison to floating-point operations. The downside to this gained efficiency is an added 2-bytes, totalling in at a requirement of 1-byte less than simply using a float. The upside is that I can scale back to near single precision accuracy well past the boundary of the functions domain while experiencing mild clipping in quadrant I.

Int8 is often used in other applications due to requirements for manually adjusting required parameters. But I suppose well see a slight shift if hardware manufacturers find value in this added complexity.

No matter which we choose, we create a trade-off between efficiency, precision, and storage.

If we look at the line you claimed was the squeezing factor and inspect it more closely, we'll see that the inverse of delta is actually the **step size** of the **domain**.

```c
const float id = d ? 1.0f/d : 0.0f; // 1 / d steps
```

The squeezing factor "squeezes" the real number into the integer domain. This is not what that is.
